presto-main/src/main/java/com/facebook/presto/dispatcher/QueryInfoFetcher.java

```diff
@@ -0,0 +1,323 @@
+
+    public void addFinalQueryInfoListener(StateMachine.StateChangeListener<BasicQueryInfo> stateChangeListener)
+    {
+        AtomicBoolean done = new AtomicBoolean();
+        StateMachine.StateChangeListener<Optional<BasicQueryInfo>> fireOnceStateChangeListener = finalQueryInfo -> {
+            if (finalQueryInfo.isPresent() && done.compareAndSet(false, true)) {
+                stateChangeListener.stateChanged(finalQueryInfo.get());
+            }
+        };
+        finalQueryInfo.addStateChangeListener(fireOnceStateChangeListener);
+        fireOnceStateChangeListener.stateChanged(finalQueryInfo.get());
+    }
+
+    private synchronized void scheduleUpdate(URI location)
+    {
+        checkState(!running, "Already scheduled");
+        scheduledFuture = updateScheduledExecutor.scheduleWithFixedDelay(() -> {
+            synchronized (this) {
+                // if the previous request still running, don't schedule a new request
+                if (future != null && !future.isDone()) {
+                    return;
+                }
+            }
+            if (nanosSince(lastUpdateNanos.get()).toMillis() >= updateIntervalMillis) {
+                sendNextRequest(location);
+            }
+        }, 0, 100, MILLISECONDS);
+    }
+
+    private synchronized void sendNextRequest(URI location)
+    {
+        QueryState queryState = getQueryInfo().getState();
+
+        if (!running) {
+            return;
+        }
+
+        // we already have the final query info
+        if (queryState.isDone()) {
+            stop();
+            return;
+        }
+
+        // if we have an outstanding request
+        if (future != null && !future.isDone()) {
+            return;
+        }
+
+        // if throttled due to error, asynchronously wait for timeout and try again
+        ListenableFuture<?> errorRateLimit = errorTracker.acquireRequestPermit();
+        if (!errorRateLimit.isDone()) {
+            errorRateLimit.addListener(() -> sendNextRequest(location), executor);
+            return;
+        }
+
+        Request.Builder requestBuilder = setContentTypeHeaders(isBinaryTransportEnabled, prepareGet());
+
+        if (queryInfoRefreshMaxWait.toMillis() != 0L) {
+            requestBuilder.setHeader(PRESTO_CURRENT_STATE, queryState.toString())
+                    .setHeader(PRESTO_MAX_WAIT, queryInfoRefreshMaxWait.toString());
+        }
+
+        Request request = requestBuilder.setUri(location).build();
+
+        ResponseHandler responseHandler;
+        if (isBinaryTransportEnabled) {
+            responseHandler = createFullSmileResponseHandler((SmileCodec<BasicQueryInfo>) queryInfoCodec);
+        }
+        else {
+            responseHandler = createAdaptingJsonResponseHandler(unwrapJsonCodec(queryInfoCodec));
+        }
+
+        errorTracker.startRequest();
+        future = httpClient.executeAsync(request, responseHandler);
+        currentRequestStartNanos.set(System.nanoTime());
+
+        Futures.addCallback(
```

Comment:
nit: static import

Reply:
I see more instances than otherwise where we don't static import this.  It seems to be the convention to avoid static importing `Futures.addCallback`.